yk_gmd_blender/yk_gmd/structs/vertex_buffer_layout.py

```diff
@@ -22,50 +18,120 @@ class VertexBufferLayoutStruct(BaseBigEndianStructure):
+                shift = (vertex_packing >> (i + 2)) & 3
+                if shift:
+                    if shift != 1:
+                        # add uv_count as index per type (highest comes first)
+                        # each 0x1 is a 4 * 8bit fixed point
+                        result += 1 + (uv_count << 4)
+                        continue
+                    # each 0x100 is a Vector2Half
+                    result += (1 << 8) + (uv_count << 12)
+                else:
+                    # each 0x10000 is a Vector2
+                    result += (1 << 16) + (uv_count << 20)
+            return result
 
     def get_vertex_layout(self) -> GMDVertexBufferLayout:
+        vertex_elems = {}
         vertex_layout_bits = [
-            ("pos", 0x03),
-            ("weights", 0x3c0),
-            ("normal", 0x1c00),
-            ("tangent", 0xe000),
-
-            ("col0", 0x00e0_0000),
-            ("col1", 0x0700_0000),
-            ("uv",   0x1800_0000),
+            ("pos", 0x07),
+            ("weights", 0x70),
+            ("bones", 0x200),
+            ("normal", 0x400),
+            ("tangent", 0x2000),
+            ("unk", 0x10000),
+            ("col0", 0x0020_0000),
+            ("col1", 0x0100_0000),
+            ("uv",   0xf000_0000),
         ]
-        overall_bitmask = 0
-        vertex_elems = {}
-        for (name, bitmask) in vertex_layout_bits:
-            overall_bitmask = overall_bitmask | bitmask
-            vertex_elems[name] = self.all_bits(int(self.vertex_packing), bitmask)
-        # overall_bitmask should equal 0x1FE0_FFC3
-        # there should also be no bits set outside of the bitmask
 
-        uv_bytes = 0
-        if self.uv_flags == 0x4:
-            uv_bytes = 2
-        elif self.uv_flags == 0x44:
-            uv_bytes = 4
+        for (name, bitmask) in vertex_layout_bits:
+            vertex_elems[name] = 0 if int(self.vertex_packing) & bitmask else -1
+        for name in vertex_elems:
+            if vertex_elems[name] == -1:
+                vertex_elems[name] = 0
+                continue
+            vertex_elems[name] = self.get_property_type(name)
 
         print(vertex_elems)
 
         return GMDVertexBufferLayout(
-            pos_en=vertex_elems["pos"],
-            weights_en=vertex_elems["weights"],
-            normal_en=vertex_elems["normal"],
-            tangent_en=vertex_elems["tangent"],
-            col0_en=vertex_elems["col0"],
-            col1_en=vertex_elems["col1"],
-            uv_en=vertex_elems["uv"],
-            uv_elem_bytes=uv_bytes
+            pos_type=vertex_elems["pos"],
+            weights_type=vertex_elems["weights"],
+            bones_en=vertex_elems["bones"],
+            normal_type=vertex_elems["normal"],
+            tangent_type=vertex_elems["tangent"],
+            unk_type=vertex_elems["unk"],
+            col0_type=vertex_elems["col0"],
+            col1_type=vertex_elems["col1"],
+            uv_type=vertex_elems["uv"],
```

Comment:
There should probably be two uv types, one for uv0, one for uv1? Unless I'm misunderstanding something

Reply:
I rewrote the format to avoid confusions. It's now uv0_type and uv1_type